src/Dictionaries/DictionaryFactory.h

```diff
@@ -55,13 +55,18 @@ class DictionaryFactory : private boost::noncopyable
 
     bool isComplex(const std::string & layout_type) const;
 
-    void registerLayout(const std::string & layout_type, LayoutCreateFunction create_layout, bool is_layout_complex);
+    /// If the argument `layout_type` is not complex layout and has corresponding complex layout,
+    /// change `layout_type` to corresponding complex and return true; otherwise do nothing and return false.
+    bool convertToComplex(std::string & layout_type) const;
+
+    void registerLayout(const std::string & layout_type, LayoutCreateFunction create_layout, bool is_layout_complex, bool has_layout_complex = true);
```

Comment:
Maybe it is a good idea not to rely on defaults, and instead, pass it explicitly everywhere, @kitaisreal what do you think?